### PR TITLE
add tcp_flags matching to forwarding module

### PR DESCRIPTION
**Pull request by allewwaly**
add tcp_flags matching to forwarding module based on work on loxigen
as discussed at https://groups.google.com/a/openflowhub.org/forum/#!topic/floodlight-dev/uSnA0IgkBJs

users can choose to match on tcp_flags or not by configuring net.floodlightcontroller.forwarding.Forwarding.match of floodlightdefault.properties (there is a flag option added)

The corresponding loxigen modification is floodlight/loxigen#512

### CODE REVIEW

File: src/main/java/net/floodlightcontroller/forwarding/Forwarding.java [hunk 1]
```diff
@@ -607,6 +608,9 @@ protected Match createMatchFromPacket(IOFSwitch sw, OFPort inPort, OFPacketIn pi
                     if (FLOWMOD_DEFAULT_MATCH_TRANSPORT_DST) {
                         mb.setExact(MatchField.TCP_DST, tcp.getDestinationPort());
                     }
+                    if(FLOWMOD_DEFAULT_MATCH_TCP_FLAG){
```
Comment (rizard) [inline src/main/java/net/floodlightcontroller/forwarding/Forwarding.java]: @allewwaly we should probably further condition this on the OpenFlow version and description of the switch. If the switch is not OpenFlow 1.2+ and is not OVS, we should not match on TCP flags.

File: src/main/java/net/floodlightcontroller/util/MatchUtils.java [hunk 2]
```diff
@@ -136,6 +136,7 @@
 	public static final String STR_PBB_UCA = "pbb_uca";
 
 	public static final String STR_TCP_FLAGS = "tcp_flags";
+	public static final String STR_OVS_TCP_FLAGS = "ovs_tcp_flags";//experimenter OXM before floodlight completely support of1.5
```
Comment (rizard) [inline src/main/java/net/floodlightcontroller/util/MatchUtils.java]: @allewwaly to be clear, Floodlight does support OpenFlow 1.5. It's OVS that experimentally supports OpenFlow 1.5

File: src/main/java/net/floodlightcontroller/forwarding/Forwarding.java [hunk 3]
```diff
@@ -653,6 +657,9 @@ protected Match createMatchFromPacket(IOFSwitch sw, OFPort inPort, OFPacketIn pi
...
```
Comment (rizard) [inline src/main/java/net/floodlightcontroller/forwarding/Forwarding.java]: @allewwaly ditto

File: src/main/java/net/floodlightcontroller/forwarding/Forwarding.java (hunk 1 above)
Comment (allewwaly) [inline src/main/java/net/floodlightcontroller/forwarding/Forwarding.java]: Agree

File: src/main/java/net/floodlightcontroller/util/MatchUtils.java (hunk 2 above)
Comment (allewwaly) [inline src/main/java/net/floodlightcontroller/util/MatchUtils.java]: I will uncomment that

File: src/main/java/net/floodlightcontroller/forwarding/Forwarding.java (hunk 1 above)
Comment (allewwaly) [inline src/main/java/net/floodlightcontroller/forwarding/Forwarding.java]: Could you please tell me how to judge switch type and ovs version in createMatchFromPacket()?

Comment (rizard) [inline src/main/java/net/floodlightcontroller/forwarding/Forwarding.java]: @allewwaly you can use IOFSwitch's getDescription() to get switch hardware and software information. I'd suggest connecting to an OVS and logging what it sends (Wireshark or OVS commands would also work). I'd then pick the minimal checks to see if a switch is OVS (and ideally the version of OVS that first supported the flags).

**Pull request closed without merging**